### ISSUE TITLE
Support for libcamera?

### ISSUE THREAD

**Issue by rustysprocket** [caprica/picam-native#4]
Do you have any plans to support libcamera? I've got some Arducam cameras I'd like to use with a Java app.

**Comment by caprica** [comment]
Does libcamera support the Raspberry Pi cameras? LibMMAL was horrible to deal with to be honest and an alternative would be nice.

The longer term plan for this project would be to drop the JNA approach it currently uses and then use Java FFI instead - but Java FFI is probably still at least two JDK releases away from being ready, and it would force users of this library onto JDK 20 or JDK 21 at a time when probably JDK 11 is where people are stuck.

If libcamera has a nice/clean/simple API I would certainly consider looking at it, I would love to forget about MMAL completely.

If libcamera doesn't support the Pi cameras, then it's a bit trickier because I don't use and don't have _any_ Arduino hardware.

**Comment by caprica** [comment]
I should add, what I'm trying to say here as well is that if I were to look at libcamera, it would be with a view to using FFI rather than JNA.

JNA is great, but for my projects it's really becoming a dead-end. For native library integration I only want to use FFI since it will be a core part of the JDK.

So if libcamera were a viable thing to use for this project, it would almost certainly be using FFI and thereby pushing you to currently a preview JDK, or maybe JDK 20 or JDK 21.

**Comment by rustysprocket** [comment]
Thanks for responding! :)

Libcamera is the Raspberry Pi's new official way of accessing cameras. It's supposed to be a more open solution and allows for more customization apparently. [Libcamera](https://www.raspberrypi.com/documentation/accessories/camera.html)

[Adrucam](https://arducam.com) makes all sorts of cameras, some for the Pi. Specifically they offer a 16MP auto focus camera as well as add on boards that allow dual or quad camera abilities. The camera drivers they provide integrate with libcamera so anything supporting libcamera *should* support their hardware.

For my project, I can control what JDK is used so requiring something new isn't the end of the world.

**Comment by caprica** [comment]
I was just reading through the libcamera API, it seems kinda big.

I'll try and find some time to look at some samples to try and gauge what sort of effort would be required.

This project got stalled because honestly I was sick of MMAL.

**Comment by rustysprocket** [comment]
When you say FFI, are you referring to this? [Panama Foreign Function and Memory API](https://github.com/openjdk/panama-foreign)

**Comment by rustysprocket** [comment]
I should add that I'm happy to try to help if it would make things easier.

I think my options are (in order of preference):
1. get some kind of Java -> libcamera library working
2. use some hacky wrapper on the libcamera apps (similar to how people wrapped raspistill)
3. reimplement the app in python

I imagine all would be time consuming but I'd rather do #1.

**Comment by caprica** [comment]
Indeed yes, Panama was the precursor to the FFI module that made its way I think initially in JDK 17, and it's available as an early access module in JDK 18.

It will be promoted to a preview in JDK 19, and after that it might be final with JDK 20 but it could be later.

The API has changed quite a bit from Panama to preview in JDK 19.

I have another project that uses this FFI module but its way too early to publish it.

Being completely honest, I would like to have an FFI solution that worked with cameras on the Pi, but right now I simply have no time to start on that at least in the short term.

**Comment by rustysprocket** [comment]
Not making any promises about how far I'll get but I've setup JDK 19 and jextract to start working through some tutorials. I'm going to try to get libcamera working. It seems jextract needs C headers and libcamera is C++ but https://foojay.io/today/category/jeps/project-panama/ has some beginner tutorials and one covers C++.

**Comment by caprica** [comment]
So all the samples I've seen so far with libcamera use it as a C++ library, i.e. you write your code in C++ and use the classes defined in libcamera.

For a foreign-language interface don't you actually need to have extern C-style methods delcared in order to be able to access those methods in a shared library?

**Comment by rustysprocket** [comment]
Yes, that's true. I'm still looking into the options but the most straight forward one may be to code up a C++ intermediate layer that exposes the necessary functions in C. I admit, it feels a bit kludgy.

**Comment by caprica** [comment]
For me, that defeats the point really - the idea is you don't want to deal with native code but now you have to write "C" shared library wrapper to a C++ library and maintain that. It just seems counter to what using Java would give you.

I'm surprised that libcamera is not a shared library in the "traditional" sense, it would be much more useful IMO if it were.

**Comment by rustysprocket** [comment]
Ok, I confess I've never done any JNI work. But after doing some reading, creating a wrapper is basically what JNI is already doing?

I do see some extern C statements in the libcamera code. I still have to dig in further so I can't say if they're exposing the necessary functions.

**Comment by rustysprocket** [comment]
After more reading I'm pretty sure libcamera does not expose anything useful via extern C. And to echo one of your previous observations, the libcamera api seems vast. The amount of work to do a wrapper or attempt to convince them to add C friendliness seems too daunting.

The one last hope might be JavaCPP. Do you have any opinions on it? Apparently it does still use JNI at the backend.

**Comment by caprica** [comment]
This is a disappointing outcome but it looks like a dead-end to me.

Never used JavaCPP, I don't know what is needed to use it.

**Comment by ajsnarr98** [comment]
It sounds like libcamera has python bindings in the works, I'm not very familiar with python-to-c bindings, but would having that change the feasibility of using it with ffi more easily in something like java?

https://www.raspberrypi.com/documentation/accessories/camera.html
> libcamera is still missing certain features, most notably Python bindings. Whilst this work is in progress, users who need a Python interface (such as Picamera) will have to stay with the legacy stack for the time being

**Comment by caprica** [comment]
Is it really going to be Python-to-C or are they going to just port their C++ code to Python?

If Python-to-C were possible, then does that not imply they have a "C" library to call into?

Another way to look at this is what does the C++ code really use under the hood, there is likely to be a C library for that. If it still uses MMAL, well, that's a shame.

In any case, my only interest in a Java API for this would be if it can use Java's FFI in the new JDKs, anything else is for some other project to pick up and run with.
